zio-http/src/main/scala/zio/http/endpoint/internal/EndpointClient.scala

```diff
@@ -41,15 +41,17 @@ private[endpoint] final case class EndpointClient[I, E, O, M <: EndpointMiddlewa
         // Preferentially decode an error from the handler, before falling back
         // to decoding the middleware error:
         val leftError =
-          endpoint.error.decodeResponse(response).orDie.flatMap((e: E) => ZIO.fail(alt.left(e)))
+          endpoint.error.decodeResponse(response).map(e => alt.left(e))
 
-        val rightError =
+        val rightError = if (invocation.middleware == EndpointMiddleware.None) {
+          ZIO.dieMessage("Middleware is none")
+        } else {
           invocation.middleware.error
             .decodeResponse(response)
-            .orDie
-            .flatMap((e: invocation.middleware.Err) => ZIO.fail(alt.right(e)))
+            .map(e => alt.right(e))
+        }
 
-        leftError.orElse(rightError)
+        leftError.orElse(rightError).orDie.flip
```

Comment:
Looks nice in general! 

I think handling unexpected error bodies could be better.

It's ok to keep the original design and we fail in the error channel with the known error messages (decoded either by the endpoint's `outErr` codecs or the middleware's error codec) and _die_ on unexpected ones.

But right now:
- if either `leftError` or `rightError` manages to decode the response, we fail with that as expected
- if both `leftError` and `rightError` fails, we die with the error from the second one
- if `leftError` fails and there is no middleware, we fail with a `Middleware is none` message

as these errors can happen in production (let's say a proxy between your service and the called one starts to respond with some unexpected rate limiting response, or just somebody deployed a new version of the target service that breaks the expected error format) it's important to have some information in the exception about what happened.

So I propose to fail with an exception which 
- contains as an inner exception the decoder failure
- if both decoder fails, contains both decoder failures
- the status code from the response
- maybe even a first N characters from the reponse body (but not sure about this one)
-

Reply:
Also we cannot simply ignore the middleware error. In fact, we need a test to make sure we properly handle middleware error.

Reply:
So I've added tests for middleware and rewrite the code a bit - I don't know how to create some kind of tree structure with Causes - that's the bast I could come up with